File: ingest_upstash.py
#!/usr/bin/env python3
"""
Ingest script for Digital Twin profile into Upstash Vector DB.
Upstash will handle embeddings automatically (configured with groq-embed-1).
"""

import os
import json
import sys
from dotenv import load_dotenv
from upstash_vector import Index

# Load environment variables
load_dotenv(dotenv_path=os.path.join(os.path.dirname(__file__), ".env.local"))

# Configuration
UPSTASH_VECTOR_URL = os.getenv("UPSTASH_VECTOR_REST_URL")
UPSTASH_VECTOR_TOKEN = os.getenv("UPSTASH_VECTOR_REST_TOKEN")

# Profile file location
profile_path = os.path.join(os.path.dirname(__file__), "digitaltwin.json")

def validate_credentials():
    """Validate that all required credentials are present."""
    if not UPSTASH_VECTOR_URL:
        print("❌ UPSTASH_VECTOR_REST_URL not found in .env.local")
        return False
    if not UPSTASH_VECTOR_TOKEN:
        print("❌ UPSTASH_VECTOR_REST_TOKEN not found in .env.local")
        return False
    print("✅ Upstash credentials found")
    return True

def load_profile():
    """Load the digital twin profile from JSON."""
    if not os.path.exists(profile_path):
        print(f"❌ Profile file not found: {profile_path}")
        return None
    
    try:
        with open(profile_path, "r", encoding="utf-8") as f:
            profile = json.load(f)
        print(f"✅ Loaded profile from {profile_path}")
        return profile
    except Exception as e:
        print(f"❌ Error loading profile: {e}")
        return None

def build_content_chunks(profile):
    """Build content chunks from the profile for ingestion."""
    chunks = []
    
    # Personal section
    personal = profile.get("personal", {})
    if personal:
        name = personal.get("name", "")
        title = personal.get("title", "")
        summary = personal.get("summary", "")
        location = personal.get("location", "")
        chunk_text = f"Personal: {name}, {title}. Location: {location}. Summary: {summary}"
        chunks.append({
            "id": "personal",
            "text": chunk_text,
            "metadata": {"type": "personal", "name": name}
        })
    
    # Experience section
    for i, exp in enumerate(profile.get("experience", [])):
        title = exp.get("title", "")
        company = exp.get("company", "")
        duration = exp.get("duration", "")
        description = exp.get("description", "")
        
        # Flatten STAR bullets
        achievements = []
        for star in exp.get("achievements_star", []):
            s = f"Situation: {star.get('situation', '')} Task: {star.get('task', '')} Action: {star.get('action', '')} Result: {star.get('result', '')}"
            achievements.append(s)
        
        chunk_text = f"Experience: {title} at {company} ({duration}). {description}. " + " ".join(achievements)
        chunks.append({
            "id": f"experience_{i}",
            "text": chunk_text,
            "metadata": {"type": "experience", "title": title, "company": company}
        })
    
    # Skills section
    skills = profile.get("skills", {})
    if skills:
        tech_skills = skills.get("technical", {})
        soft_skills = skills.get("soft_skills", [])
        
        tech_text = ", ".join([f"{k}: {v}" for k, v in tech_skills.items()])
        soft_text = ", ".join(soft_skills)
        
        chunk_text = f"Technical Skills: {tech_text}. Soft Skills: {soft_text}"
        chunks.append({
            "id": "skills",
            "text": chunk_text,
            "metadata": {"type": "skills"}
        })
    
    # Projects section
    for i, proj in enumerate(profile.get("projects_portfolio", [])):
        name = proj.get("name", "")
        description = proj.get("description", "")
        technologies = ", ".join(proj.get("technologies", []))
        impact = proj.get("impact", "")
        
        chunk_text = f"Project: {name}. Description: {description}. Technologies: {technologies}. Impact: {impact}"
        chunks.append({
            "id": f"project_{i}",
            "text": chunk_text,
            "metadata": {"type": "project", "name": name}
        })
    
    print(f"✅ Built {len(chunks)} content chunks")
    return chunks

def embed_chunks_with_groq(chunks):
    """Embed chunks using Groq's groq-embed-1 model."""
    print(f"🔄 Embedding {len(chunks)} chunks with Groq (groq-embed-1)...")
    
    groq_client = Groq(api_key=GROQ_API_KEY)
    embedded_chunks = []
    
    for i, chunk in enumerate(chunks):
        try:
            # Call Groq embedding endpoint
            response = groq_client.embeddings.create(
                input=chunk["text"],
                model="text-embedding-3-small"  # Using text-embedding-3-small as groq-embed-1 may not be available
            )
            embedding = response.data[0].embedding
            
            embedded_chunks.append({
                "id": chunk["id"],
                "values": embedding,
                "metadata": chunk["metadata"]
            })
            
            if (i + 1) % 5 == 0:
                print(f"  ✓ Embedded {i + 1}/{len(chunks)} chunks")
        except Exception as e:
            print(f"❌ Error embedding chunk {chunk['id']}: {e}")
            return None
    
    print(f"✅ Embedded all {len(embedded_chunks)} chunks")
    return embedded_chunks

def upload_to_upstash(chunks):
    """Upload text chunks to Upstash Vector DB (Upstash handles embedding automatically)."""
    print(f"🔄 Uploading {len(chunks)} text chunks to Upstash...")
    print("   (Upstash will automatically embed them using groq-embed-1)")
    
    try:
        index = Index(url=UPSTASH_VECTOR_URL, token=UPSTASH_VECTOR_TOKEN)
        
        # Upsert text directly - Upstash will embed automatically
        vectors = [
            (chunk["id"], chunk["text"], chunk["metadata"])
            for chunk in chunks
        ]
        
        index.upsert(vectors=vectors)
        print(f"✅ Successfully uploaded {len(vectors)} text chunks to Upstash")
        print("   Upstash is now embedding them with groq-embed-1...")
        
        # Verify the upload
        try:
            info = index.info()
            vector_count = info.get("vector_count", 0) if isinstance(info, dict) else getattr(info, "vector_count", 0)
            print(f"📊 Total vectors in index: {vector_count}")
        except Exception as e:
            print(f"⚠️  Could not fetch index info: {e}")
        
        return True
    except Exception as e:
        print(f"❌ Error uploading to Upstash: {e}")
        if "embedding" in str(e).lower():
            print("\n⚠️  IMPORTANT: Your Upstash Vector index may not be configured with an embedding model.")
            print("   Please recreate your index with 'groq-embed-1' or another embedding model:")
            print("   1. Go to https://console.upstash.com")
            print("   2. Delete the current Vector index")
            print("   3. Create a new index with Embedding Model: 'groq-embed-1'")
            print("   4. Copy the new REST URL and Token to .env.local")
            print("   5. Run this script again")
        return False

def main():
    """Main ingestion workflow."""
    print("🚀 Digital Twin Upstash Ingestion Script")
    print("=" * 50)
    
    # Step 1: Validate credentials
    if not validate_credentials():
        sys.exit(1)
    
    # Step 2: Load profile
    profile = load_profile()
    if not profile:
        sys.exit(1)
    
    # Step 3: Build content chunks
    chunks = build_content_chunks(profile)
    if not chunks:
        print("❌ No content chunks generated")
        sys.exit(1)
    
    # Step 4: Upload to Upstash (Upstash handles embedding)
    if upload_to_upstash(chunks):
        print("\n✅ Ingestion complete! Your Digital Twin profile is now searchable.")
    else:
        sys.exit(1)

if __name__ == "__main__":
    main()

File: scripts/ingest_upstash.js
#!/usr/bin/env node
/*
  ingest_upstash.js

  CLI to create an Upstash Vector index (optionally) and ingest a flattened
  `digitaltwin.json` profile into it. Prompts for embedding provider and model.

  Usage (PowerShell):
    node .\mydigitaltwin\scripts\ingest_upstash.js

  Env vars expected (one of the patterns):
    UPSTASH_VECTOR_REST_URL, UPSTASH_VECTOR_REST_TOKEN
    or NEXT_PUBLIC_UPSTASH_VECTOR_REST_URL, NEXT_PUBLIC_UPSTASH_VECTOR_REST_TOKEN

  Embedding providers supported:
    - openai: requires OPENAI_API_KEY
    - groq: requires GROQ_API_KEY and optional GROQ_API_URL

  Notes:
    - The script will attempt to create an index via the Upstash REST `indexes`
      endpoint if you choose to. If creation fails, you'll be prompted to create
      the index manually in the Upstash console (or run this script again after fixing credentials).
    - Upstash may reject upserts if the index was not created with an embedding model.
*/

const fs = require("fs")
const path = require("path")
const readline = require("readline")

// Try to load dotenv from common locations without overwriting existing envs
try {
  // Use dynamic require to avoid bundlers attempting to resolve dotenv at build time
  const dynamicRequire = eval("require")
  const dotenv = dynamicRequire ? dynamicRequire("dotenv") : null
  const cwd = process.cwd()
  const candidates = [
    path.join(cwd, ".env.local"),
    path.join(cwd, ".env"),
    path.join(cwd, "../.env.local"),
    path.join(cwd, "../.env"),
  ]
  for (const p of candidates) {
    try {
      if (fs.existsSync(p)) {
        if (dotenv && typeof dotenv.config === "function") dotenv.config({ path: p })
        break
      }
    } catch (e) {}
  }
} catch (e) {}

const { Index } = (() => {
  try {
    return require("@upstash/vector")
  } catch (e) {
    return { Index: null }
  }
})()

function log(...args) {
  console.log("[ingest]", ...args)
}
function error(...args) {
  console.error("[ingest][error]", ...args)
}

async function prompt(question) {
  const rl = readline.createInterface({ input: process.stdin, output: process.stdout })
  return new Promise((resolve) => rl.question(question, (ans) => { rl.close(); resolve(ans) }))
}

function findProfile() {
  const candidates = [
    path.join(process.cwd(), "data", "digitaltwin.json"),
    path.join(process.cwd(), "digitaltwin.json"),
    path.join(process.cwd(), "mydigitaltwin", "data", "digitaltwin.json"),
    path.join(process.cwd(), "mydigitaltwin", "digitaltwin.json"),
    path.join(__dirname, "..", "data", "digitaltwin.json"),
  ]
  for (const p of candidates) {
    if (fs.existsSync(p)) return p
  }
  return null
}

function buildContextFromProfile(profile) {
  const chunks = []
  const personal = profile?.personal ?? {}
  const personalText = `${personal?.name ?? ""}, ${personal?.title ?? ""}. ${personal?.summary ?? ""}`.trim()
  if (personalText) chunks.push({ id: "personal", title: "Personal Info", content: personalText, type: "personal" })

  const experiences = profile?.experience ?? []
  for (let i = 0; i < experiences.length; i++) {
    const exp = experiences[i]
    let expText = `${exp?.title ?? ""} at ${exp?.company ?? ""} (${exp?.duration ?? ""}). `
    for (const ach of (exp?.achievements_star ?? [])) {
      expText += `Situation: ${ach?.situation ?? ""}. Task: ${ach?.task ?? ""}. Action: ${ach?.action ?? ""}. Result: ${ach?.result ?? ""}. `
    }
    chunks.push({ id: `exp_${i}`, title: exp?.title ?? "", content: expText.trim(), type: "experience" })
  }

  const skills = profile?.skills ?? {}
  const tech = skills?.technical ?? {}
  const techStr = Object.entries(tech).map(([k, v]) => `${k}: ${JSON.stringify(v)}`).join(", ")
  const soft = (skills?.soft_skills ?? []).join(", ")
  const skillText = `Technical skills: ${techStr}. Soft skills: ${soft}`.trim()
  if (skillText) chunks.push({ id: "skills", title: "Skills", content: skillText, type: "skills" })

  const projects = profile?.projects_portfolio ?? []
  for (let i = 0; i < projects.length; i++) {
    const proj = projects[i]
    const projText = `${proj?.name ?? ""}: ${proj?.description ?? ""}. Technologies: ${(proj?.technologies ?? []).join(", ")}. Impact: ${proj?.impact ?? ""}.`.trim()
    chunks.push({ id: `proj_${i}`, title: proj?.name ?? "", content: projText, type: "project" })
  }

  return chunks
}

async function embedTextOpenAI(model, texts) {
  const key = process.env.OPENAI_API_KEY || process.env.NEXT_PUBLIC_OPENAI_API_KEY
  if (!key) throw new Error("OPENAI_API_KEY not set in environment")
  const url = "https://api.openai.com/v1/embeddings"
  const body = { model, input: texts }
  const resp = await fetch(url, {
    method: "POST",
    headers: { "Content-Type": "application/json", Authorization: `Bearer ${key}` },
    body: JSON.stringify(body),
  })
  if (!resp.ok) throw new Error(`OpenAI embeddings request failed: ${resp.status} ${await resp.text()}`)
  const j = await resp.json()
  // j.data is array of { embedding }
  return j.data.map((d) => d.embedding)
}

async function embedTextGroq(model, texts) {
  const key = process.env.GROQ_API_KEY || process.env.NEXT_PUBLIC_GROQ_API_KEY
  const apiUrl = process.env.GROQ_API_URL || process.env.NEXT_PUBLIC_GROQ_API_URL || "https://api.groq.ai/v1"
  if (!key) throw new Error("GROQ_API_KEY not set in environment")
  const url = `${apiUrl}/embeddings`
  // attempt batching
  const resp = await fetch(url, {
    method: "POST",
    headers: { "Content-Type": "application/json", Authorization: `Bearer ${key}` },
    body: JSON.stringify({ model, input: texts }),
  })
  if (!resp.ok) throw new Error(`Groq embeddings request failed: ${resp.status} ${await resp.text()}`)
  const j = await resp.json()
  // assume j.data array like OpenAI
  return j.data.map((d) => d.embedding)
}

async function ensureIndexExists(restUrl, token, indexName, modelName) {
  // Ask user if they want us to try creating index automatically
  const yn = (await prompt(`Attempt to create index "${indexName}" on Upstash automatically? (y/N): `)).trim().toLowerCase()
  if (yn !== "y") return { created: false }

  const createUrl = `${restUrl.replace(/\/$/, "")}/indexes`
  const body = { name: indexName, embedding: { model: modelName } }
  log(`Creating index via ${createUrl} ...`)
  try {
    const resp = await fetch(createUrl, {
      method: "POST",
      headers: { "Content-Type": "application/json", Authorization: `Bearer ${token}` },
      body: JSON.stringify(body),
    })
    const text = await resp.text()
    if (!resp.ok) {
      error("Index creation failed:", resp.status, text)
      return { created: false, error: text }
    }
    log("Index created (response):", text.substring(0, 1000))
    return { created: true }
  } catch (err) {
    error("Index creation request failed:", err?.message ?? err)
    return { created: false, error: String(err) }
  }
}

async function upsertVectors(index, vectors) {
  // Try a few likely SDK shapes
  try {
    if (!index) throw new Error("Index instance missing")
    if (typeof index.upsert === "function") {
      log("Using SDK upsert(items)")
      // SDK may expect { items: [...] }
      const tryBody = { items: vectors }
      const res = await index.upsert(tryBody)
      return res
    }
    if (typeof index.bulk === "function") {
      log("Using SDK bulk()")
      const res = await index.bulk(vectors)
      return res
    }
    // otherwise try a generic request
    if (typeof index.request === "function") {
      log("Using SDK request() to call upsert")
      const res = await index.request("/upsert", { method: "POST", json: { items: vectors } })
      return res
    }
    throw new Error("No known upsert method on Index SDK instance")
  } catch (err) {
    throw err
  }
}

async function main() {
  log("Starting ingestion script")

  const profilePath = findProfile()
  if (!profilePath) {
    error("digitaltwin.json not found. Put it in data/digitaltwin.json or repo root.")
    process.exit(1)
  }
  log("Using profile:", profilePath)
  const profile = JSON.parse(fs.readFileSync(profilePath, "utf8"))
  const chunks = buildContextFromProfile(profile)
  if (!chunks.length) {
    error("No chunks created from profile — aborting")
    process.exit(1)
  }
  log(`Created ${chunks.length} chunks from profile`)

  const url = process.env.UPSTASH_VECTOR_REST_URL || process.env.NEXT_PUBLIC_UPSTASH_VECTOR_REST_URL
  const token = process.env.UPSTASH_VECTOR_REST_TOKEN || process.env.NEXT_PUBLIC_UPSTASH_VECTOR_REST_TOKEN
  if (!url || !token) {
    error("UPSTASH_VECTOR_REST_URL and UPSTASH_VECTOR_REST_TOKEN must be set in your environment")
    process.exit(1)
  }

  const indexName = (await prompt("Index name to use (default: digitaltwin): ")) || "digitaltwin"

  const provider = (await prompt("Embedding provider ('openai' or 'groq', default 'groq'): ")) || "groq"
  const model = (await prompt("Embedding model name (e.g. 'text-embedding-3-small' or 'groq-embed-1', enter model name): "))
  if (!model) {
    error("Embedding model required — aborting")
    process.exit(1)
  }

  // Optionally attempt to create the index via REST
  await ensureIndexExists(url, token, indexName, model)

  // Generate embeddings
  const texts = chunks.map((c) => c.content)
  log("Generating embeddings for chunks... provider=", provider, "model=", model)
  let embeddings = []
  try {
    if (provider === "openai") embeddings = await embedTextOpenAI(model, texts)
    else embeddings = await embedTextGroq(model, texts)
  } catch (err) {
    error("Embedding generation failed:", err?.message ?? err)
    process.exit(1)
  }

  if (!embeddings || embeddings.length !== texts.length) {
    error("Embedding length mismatch or empty embeddings returned")
    process.exit(1)
  }

  const vectors = chunks.map((c, i) => ({ id: `${indexName}:${c.id}`, vector: embeddings[i], metadata: { title: c.title, type: c.type, content: c.content } }))

  // Instantiate SDK Index and attempt upsert
  let indexInstance = null
  if (Index && Index !== null) {
    try {
      indexInstance = new Index({ url, token })
    } catch (e) {
      log("Failed to construct Index SDK; will try direct REST upsert", e?.message ?? e)
      indexInstance = null
    }
  }

  try {
    if (indexInstance) {
      const resp = await upsertVectors(indexInstance, vectors)
      log("Upsert SDK response:", JSON.stringify(resp).slice(0, 1000))
      log("Ingestion finished — vectors uploaded via SDK")
      return
    }

    // Fallback: try REST upsert endpoint. Upstash Vector REST APIs vary by deployment —
    // we try a best-effort pattern: POST { items: [...] } to `${url}/vectors` or `${url}/indexes/${indexName}/upsert`
    const candidates = [
      `${url.replace(/\/$/, "")}/vectors/upsert`,
      `${url.replace(/\/$/, "")}/indexes`,
      `${url.replace(/\/$/, "")}/indexes/${indexName}/upsert`,
      `${url.replace(/\/$/, "")}/upsert`,
    ]
    let done = false
    for (const c of candidates) {
      try {
        log("Trying REST upsert to", c)
        const r = await fetch(c, {
          method: "POST",
          headers: { "Content-Type": "application/json", Authorization: `Bearer ${token}` },
          body: JSON.stringify({ items: vectors }),
        })
        const t = await r.text()
        if (r.ok) {
          log("REST upsert succeeded to", c, "response:", t.slice(0, 1000))
          done = true
          break
        } else {
          log("REST upsert returned non-OK for", c, r.status, t.slice(0, 400))
        }
      } catch (err) {
        log("REST upsert failed for", c, err?.message ?? err)
      }
    }
    if (!done) {
      error("All upsert attempts failed. Common cause: index must be created with an embedding model. Please create an index in the Upstash console with embeddings enabled, then re-run this script.")
      process.exit(1)
    }
    log("Ingestion finished (REST fallback)")
  } catch (err) {
    error("Upsert failed:", err?.message ?? err)
    process.exit(1)
  }
}

main().catch((err) => { error(err?.stack ?? err); process.exit(1) })
